### ISSUE TITLE
Slow IntelliSense in 1.57.0

### ISSUE THREAD

**Issue by PawelJ-PL** [microsoft/TypeScript#44851]
Does this issue occur when all extensions are disabled?: **Yes**

- VS Code Version: **1.57.0-1623259737**
- OS Version: **Debian GNU/Linux 11 (bullseye)**

In **1.57.0-1623259737** version IntelliSense is terribly slow. I've no Idea, what more details should I provide. I've tested it on TypeScript/React project (also with all extensions disabled - with the same result - loading hint takes more than 10 seconds).

After downgrade to **1.56.2-1620838498** everything works great.

**Comment by joniaiuser** [comment]
> Update: this may be related to TypeScript 4.3? After updating our codebase to TS43 VSCode's Intellisense has slowed down to a crawl even when using 1.56.2.

I have this same issue, noticed VS code slowing down significantly. Downgrading to Typescript 4.2 seems to speed things up by a factor of 10.

**Comment by armanio123** [comment]
The slowness is caused by PR #30639. I'm still looking at options on how to avoid this.

**Comment by armanio123** [comment]
Alright I've been looking into it and seems like it's going to be complicated to optimize. All the problem is that typescript is doing a bit more work to relate conditional types (in this example the problem is with type `UncheckedJSONSchemaType` on file `node_modules/ajv/dist/types/json-schema.d.ts`)

Nevertheless, we can still get into a good performance by dividing the type into sub named types.  I've done some tests and manage to get `encodedSemanticClassifications-full` to about a second just by doing that.

Check the gist with how the file look after testing: https://gist.github.com/armanio123/95a33f9ad3286a7d78590eb11af293c4

I don't know much about the project, so all my types on the example are named just an alphabet letter.

**Comment by joniaiuser** [comment]
I'm not using Ajv in my system though. If it's a widespread issue, should it be highlighted as a breaking change so that library maintainers can update their types accordingly?

**Comment by PawelJ-PL** [comment]
I have reported this issue, but I also do not use `ajv`

**Comment by outranker** [comment]
> I have reported this issue, but I also do not use `ajv`

Is it possible a package in your project has ajv as a dependency? Coule you please confirm by looking at package-lock.json or yarn.lock?

**Comment by PawelJ-PL** [comment]
It appears that `ajv` actually occurs as a transitive dependency:

```
├─┬ @typescript-eslint/parser@4.26.1
│ └─┬ eslint@7.28.0
│   ├─┬ @eslint/eslintrc@0.4.2
│   │ └── ajv@6.12.6 deduped
│   ├── ajv@6.12.6
│   └─┬ table@6.7.1
│     └── ajv@8.6.0
└─┬ react-scripts@4.0.3
  ├─┬ @pmmmwh/react-refresh-webpack-plugin@0.4.3
  │ └─┬ schema-utils@2.7.1
  │   └── ajv@6.12.6 deduped
  ├─┬ eslint-webpack-plugin@2.5.4
  │ └─┬ schema-utils@3.0.0
  │   └── ajv@6.12.6 deduped
  ├─┬ file-loader@6.1.1
  │ └─┬ schema-utils@3.0.0
  │   └── ajv@6.12.6 deduped
  ├─┬ mini-css-extract-plugin@0.11.3
  │ └─┬ schema-utils@1.0.0
  │   ├─┬ ajv-errors@1.0.1
  │   │ └── ajv@6.12.6 deduped
  │   └── ajv@6.12.6 deduped
  ├─┬ postcss-loader@3.0.0
  │ └─┬ schema-utils@1.0.0
  │   └── ajv@6.12.6 deduped
  ├─┬ sass-loader@10.2.0
  │ └─┬ schema-utils@3.0.0
  │   └── ajv@6.12.6 deduped
  ├─┬ terser-webpack-plugin@4.2.3
  │ └─┬ schema-utils@3.0.0
  │   └── ajv@6.12.6 deduped
  ├─┬ url-loader@4.1.1
  │ └─┬ schema-utils@3.0.0
  │   └── ajv@6.12.6 deduped
  ├─┬ webpack-dev-server@3.11.1
  │ └─┬ schema-utils@1.0.0
  │   └── ajv@6.12.6 deduped
  └─┬ webpack@4.44.2
    ├─┬ ajv-keywords@3.5.2
    │ └── ajv@6.12.6 deduped
    ├── ajv@6.12.6 deduped
    └─┬ schema-utils@1.0.0
      └── ajv@6.12.6 deduped
```

I suspect then that here `eslint` may be the prime suspect. However, as far as I understand, I can fix in my code.

**Comment by VictorQueiroz** [comment]
Repeating my report about this here since #37746 was recently closed.

---

I'm trying to move on from 3.8.3 for a long time, because it's the only version which does not kill my productivity. I have an extremely slow code completion in anything higher than 3.8.3. Currently the logs are from TS 4.4.2. 

Sometimes resting the cursor over a property/method completely stucks TS server for seconds:

```
Perf 566  [01:39:04.841] 14::encodedSemanticClassifications-full: elapsed time (in milliseconds) 6612.9189
```

This is code completion when writing a dot after an object, after ~8 seconds I could finally see the property suggestion list:

```
Info 798  [02:05:51.334] Starting updateGraphWorker: Project: /home/work/Projects/hidden/core/tsconfig.json
Info 799  [02:05:51.380] Finishing updateGraphWorker: Project: /home/work/Projects/hidden/core/tsconfig.json Version: 3 structureChanged: false structureIsReused:: Completely Elapsed: 45.291651997715235ms
Info 800  [02:05:51.380] Different program with same set of files
Info 801  [02:05:51.382] getCompletionData: Get current token: 0.027127999812364578
Info 802  [02:05:51.382] getCompletionData: Is inside comment: 0.06167500093579292
Info 803  [02:05:51.382] getCompletionData: Get previous token: 0.03589100018143654
Info 804  [02:05:51.382] getCompletionsAtPosition: isCompletionListBlocker: 0.17612700164318085
Info 805  [02:05:55.188] getCompletionData: Semantic work: 3805.772270999849
Info 806  [02:05:55.189] getCompletionsAtPosition: getCompletionEntriesFromSymbols: 0.7570599988102913
Info 807  [02:05:55.639] IntelliCode plugin: successfully loaded model from /home/work/.vscode/extensions/visualstudioexptteam.vscodeintellicode-1.2.14/cache/1A3F919CD5B7725108B03C349630A1C82B03_4A84987F2428442A9EC84563BB2BC063
Info 808  [02:05:55.640] IntelliCode plugin: program found
Info 809  [02:05:55.640] IntelliCode plugin: source file found
Info 810  [02:05:55.641] IntelliCode plugin: determined type: "reducer".IAuthenticationForgotPasswordState
Info 811  [02:05:55.642] IntelliCode plugin: pushing prior access: email
Info 812  [02:05:55.653] IntelliCode plugin: prior accesses: email
Info 813  [02:05:55.653] IntelliCode plugin: sequences: N~email,N
Info 814  [02:05:55.653] IntelliCode plugin: sequences: N~email,N
Info 815  [02:05:55.653] IntelliCode plugin: added metadata {"Id":1630364755653,"ModelVersion":"typescript_1.5","Count":0,"FailureReason":"NotInModel","Class":"","InIf":false,"ElapsedTime":463.245825,"Methods":[],"PrevInvoc":"","ItemMetadata":{}} to completion result
Perf 816  [02:05:55.654] 95::completionInfo: elapsed time (in milliseconds) 4319.9144
```

Another test with VS Code Insiders:

```
Info 603  [02:19:20.988] Starting updateGraphWorker: Project: /home/work/Projects/hidden/core/tsconfig.json
Info 604  [02:19:21.034] Finishing updateGraphWorker: Project: /home/work/Projects/hidden/core/tsconfig.json Version: 3 structureChanged: false structureIsReused:: Completely Elapsed: 45.792923998087645ms
Info 605  [02:19:21.034] Different program with same set of files
Info 606  [02:19:21.035] getCompletionData: Get current token: 0.039069999009370804
Info 607  [02:19:21.035] getCompletionData: Is inside comment: 0.05014200136065483
Info 608  [02:19:21.035] getCompletionData: Get previous token: 0.04182099923491478
Info 609  [02:19:21.035] getCompletionsAtPosition: isCompletionListBlocker: 0.00338900089263916
Info 610  [02:19:25.077] getCompletionData: Semantic work: 4042.3234740011394
Info 611  [02:19:25.080] getCompletionsAtPosition: getCompletionEntriesFromSymbols: 1.3409719988703728
Perf 612  [02:19:25.081] 40::completionInfo: elapsed time (in milliseconds) 4093.2213
Info 613  [02:19:25.081] response:
    {"seq":0,"type":"response","command":"completionInfo","request_seq":40,"success":true,"performanceData":{"updateGraphDurationMs":45.792923998087645},"body":{"isGlobalCompletion":false,"isMemberCompletion":true,"isNewIdentifierLocation":false,"entries":[{"name":"email","kind":"property","kindModifiers":"","sortText":"11"},{"name":"error","kind":"property","kindModifiers":"optional","sortText":"11"},{"name":"isLoading","kind":"property","kindModifiers":"","sortText":"11"},{"name":"state","kind":"property","kindModifiers":"","sortText":"11"}]}}
Info 614  [02:19:25.082] request:
    {"seq":41,"type":"request","command":"encodedSemanticClassifications-full","arguments":{"file":"/home/work/Projects/hidden/core/authentication/effects/index.ts","start":2441,"length":3413,"format":"2020"}}
Perf 615  [02:19:27.203] 41::encodedSemanticClassifications-full: elapsed time (in milliseconds) 2121.3128
```

VS Code Insiders version:

```
Version: 1.60.0-insider
Commit: 44a4b19a910a287adef4daaeabcf68f2ba893c35
Date: 2021-08-30T11:36:28.164Z
Electron: 13.1.8
Chrome: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Linux x64 5.10.61-1-lts snap
```

VS Code version:

```
Version: 1.59.1
Commit: 3866c3553be8b268c8a7f8c0482c0c0177aa8bfa
Date: 2021-08-19T11:54:37.513Z
Electron: 13.1.7
Chrome: 91.0.4472.124
Node.js: 14.16.0
V8: 9.1.269.36-electron.0
OS: Linux x64 5.10.61-1-lts snap
```

My main question is, how can I find out exactly what are the expressions in this project (70k+ LoC), that are creating these problems? Any help is much appreciated. Thank you!

**Comment by andrewbranch** [comment]
1. Can anyone share a real project that’s having significant perf issues that I can use to investigate?
2. @gravyboat if you disable semantic highlighting in your VS Code settings, do other editor features become reasonably usable or are they still slow?

**Comment by gravyboat** [comment]
@andrewbranch I did go through the performance wiki (noted this in my original comment but should have clarified more) including a couple of hours trying various things in https://github.com/microsoft/TypeScript/wiki/Performance#investigating-issues before I posted originally. What does the team consider to be a large project in terms of files, library lines, definition lines, etc? I don't consider this project to be especially large, but maybe I'm wrong there.

Since text can be somewhat unclear I want to clarify I'm not questioning anything you've said or trying to argue about it (I agree with your comments). I've never encountered this problem before in any TS project so this sort of performance issue and the various components involved are totally new to me.

> I’m hereby declaring this issue about perf issues with encodedSemanticClassifications.#46228 

👍 Sounds good!

**Comment by andrewbranch** [comment]
> What does the team consider to be a large project in terms of files, library lines, definition lines, etc?

I don’t have a great answer for this off the top of my head, but I think it might be illuminating to see the output of a `tsc --extendedDiagnostics`, since that gives us all of that information and more.

**Comment by outranker** [comment]
> I'm trying to move on from 3.8.3 for a long time, because it's the only version which does not kill my productivity. I have an extremely slow code completion in anything higher than 3.8.3. Currently the logs are from TS 4.4.2.

Thank you for the info! I have downgraded to 3.8.3 and there is a huge increase in autocomplete speed. I was thinking of  moving back to JS because for the past couple of months dev exp with TS has been unbearable. I will test if this perf increase with 3.8.3 is long-lasting if so I might as well get stuck with it until the problem is solved

**Comment by mjbvz** [comment]
Instead of downgrading, you can disable semantic highlighting in VS Code by setting `"editor.semanticHighlighting.enabled": false`

**Comment by joniaiuser** [comment]
> @andrewbranch I did go through the performance wiki (noted this in my original comment but should have clarified more) including a couple of hours trying various things in https://github.com/microsoft/TypeScript/wiki/Performance#investigating-issues before I posted originally. What does the team consider to be a large project in terms of files, library lines, definition lines, etc? I don't consider this project to be especially large, but maybe I'm wrong there. I'll see about running through the performance tracing again, maybe something will show up that can help.
> 
> 
> 
> Since text can be somewhat unclear I want to clarify I'm not questioning anything you've said or trying to argue about it (I agree with your comments). I've never encountered this problem before in any TS project so this sort of performance issue and the various components involved are totally new to me.
> 
> 
> 
> > I’m hereby declaring this issue about perf issues with encodedSemanticClassifications.#46228 
> 
> 
> 
> 👍 Sounds good!

I think encodedSemanticClassifications is only part of the prop me since it only solves around 1/3 of the latency issue.

**Comment by joniaiuser** [comment]
Why is this closed? It's not fixed even in latest version. I still see significant performance regression.

**Comment by hanayashiki** [comment]
TypeScript: 4.5.4

It takes about 6-7 seconds after my identifier is correctly marked not found and another 6-7 seconds to erase the red line after I correct it.

https://youtube.com/shorts/dLvwjE8zEe8?feature=share

This is a moderately sized project (50000 lines) with 20-30 imports and the file is around 200 lines.

**Comment by nycgavin** [comment]
I'm using react-hook-form as well, and I'm also having this issue, upgrading to typescript 4.5 doesn't fix the issue

**Comment by andrewbranch** [comment]
Please file a new issue with detailed reproduction steps, [a TS Server log](https://github.com/microsoft/TypeScript/wiki/Getting-logs-from-TS-Server-in-VS-Code), and comparison with a recent version of TS that’s faster.

**Comment by DNoel26** [comment]
May help someone. Added the below to my package.json 

```
"scripts": {
        ......
        "typecheck": "tsc",
        "typecheck:watch": "tsc -w"
    },
```
and included the below in my settings.json

```
"javascript.validate.enable": false,
"typescript.validate.enable": false,
```

I use watch in a split terminal for type checking while I code. A bit less convenient but it's basically instant now vs 10-30+ seconds for something such as renaming a symbol or importing

**Comment by robertwt7** [comment]
Hey team, 

I'm also experiencing this in the latest one. I'm not sure why it's closed? Doesn't happen on my mac though. It only happens on my PC (windows).

When i press save, it takes 4-5s for typescript to update the errors in my vscode

**Comment by nickwang0808** [comment]
I'm experiencing the same issue.

Opening up Htop it seems like only 1 processor is being used to 100% while other processors are just idling.

Any ideas? 

I'm on WSL2

**Comment by nickwang0808** [comment]
> language service

Hmm, I'm trying to figure out a way to repro it, it has to be a fairly large repo.
Just out of curiosity, where can I find more info about TS language service being only single-threaded? I know node is a single thread but it also uses worker threads no?

**Comment by amcasey** [comment]
@nickwang0808 I'm not aware of a comprehensive write-up but the fundamental problem is that coordinating worker threads requires async (or an equivalent mechanism), which would spread virally through the codebase (there's no escape hatch where you can just say "wait for the async thing" and go back to being synchronous), breaking all of our public APIs.  It would also require substantial re-architecture, since there's a lot of mutation (for perf reasons) that wouldn't play nicely with multiple threads.

TL;DR: We'd have to make an enormous, API-breaking code change and getting back to the current level of performance would be challenging

**Comment by nickwang0808** [comment]
> @nickwang0808 I'm not aware of a comprehensive write-up but the fundamental problem is that coordinating worker threads requires async (or an equivalent mechanism), which would spread virally through the codebase (there's no escape hatch where you can just say "wait for the async thing" and go back to being synchronous), breaking all of our public APIs. It would also require substantial re-architecture, since there's a lot of mutation (for perf reasons) that wouldn't play nicely with multiple threads.
> 
> TL;DR: We'd have to make an enormous, API-breaking code change and getting back to the current level of performance would be challenging

OK so I just ran a test on a macos machine with same repo. autocomplete uses all thread, where as my machine windows-wsl2 uses only 1 thread. It was instant autocomplete on the macos where mine was around 3 sec.

**Comment by amcasey** [comment]
The number of threads is the same on every platform.  It is, however, true that there is a helper process that will provide approximate answers until the language service is finished initializing, so it's possible that you're seeing that helper process consume a core and provide a quick answer.  In some circumstances, there may also be helper processes to download type definitions automatically, but the language service proper will still use only one core.

**Comment by nickwang0808** [comment]
> The number of threads is the same on every platform. It is, however, true that there is a helper process that will provide approximate answers until the language service is finished initializing, so it's possible that you're seeing that helper process consume a core and provide a quick answer. In some circumstances, there may also be helper processes to download type definitions automatically, but the language service proper will still use only one core.

Here is a side by side of native windows vs WSL2
https://user-images.githubusercontent.com/63834422/169404473-76d513a1-d123-4623-8885-7c434b8d4b2b.mp4

https://user-images.githubusercontent.com/63834422/169404620-fc863967-a8c2-416c-8d54-9c750947a399.mp4

So I'm not sure where to look, can you point me to the right direction?

**Comment by amcasey** [comment]
> So I'm not sure where to look, can you point me to the right direction?

Can you elaborate?  I'm not sure what you're looking for.

If your basic question is "why is the language service slower on WSL2 than on the host windows machine?", some possible causes include:
- The WSL2 VM not being able to use all cores (I'm not sure what settings control this)
- The WSL2 VM not being able to use enough memory (node/electron bogs down doing repeated GCs when it's near its limit)
- The fact that Linux doesn't support recursive file watches, so the LS has to use polling
- Accessing files across the Windows/Linux boundary (e.g. if you're binaries or sources are on your Windows file system)

It might also be worthwhile to determine which processes are using your cores on Windows because they can't all be the LS process.

**Comment by sannajammeh** [comment]
Getting the same issues here.  150+ file project.

Weirdly enough its only on Windows. I have an M1 Mac which receives sub 1s Typescript results yet my Windows PC (8 core, 3GHz) takes 30 seconds for almost every change.

**Comment by arnm** [comment]
Using combination of prisma, trpc, zod. About 12 prisma models so I guess its a lot of generated code. TS version 4 and 5 destroy my CPU with constant 100% utilization. If I downgrade to TS version 3.8.3 the issue goes away but I lose all the type inference :/ Please fix performance issue

**Comment by TheRealFlyingCoder** [comment]
Also experiencing these performance issues on Windows.

Wondering why this was closed without any real resolution?

Happy to run specific tests, but I am also using Prisma and Zod on a large monorepo

**Comment by VictorQueiroz** [comment]
@TheRealFlyingCoder Windows is not eligible for any sort of development.

**Comment by mjbvz** [comment]
@TheRealFlyingCoder (and others) If you are still seeing performance problems, please open a new issue against TypeScript with an example project which demonstrates the issue. Thanks
